File: CA/Projeto_1/C++/encrypt.cpp
#include "Sboxes.h"


int main(int argc, char* argv[]){

    if (argc < 2 || argc > 3) {
        std::cout << "Usage: ./encrypt <password> <mode:optional>" << std::endl;
        return 1;
    }

    std::string key_string = argv[1];
    std::string mode = "des";

    if (argc == 3) {
        mode = argv[2];
    }

    std::string text;
    std::getline(std::cin, text);
    std::vector<unsigned char> bytes(text.begin(), text.end());
    bytes.push_back('\n');

    //Padder
    bytes = addPaddingPKCS7(bytes);

    std::vector<unsigned char> encrypted_text;
    if (mode == "edes")
    {   
        //Sbox Generation
        Sboxes box = Sboxes::sbox_gen(key_string);
        //Encrypt Edes
        encrypted_text = encryptEDES(bytes, box);
    }
    else
    {   
        //Key Generation
        std::array<unsigned char, 8> key = generate_key(key_string);
        DES_key_schedule keysched;
        DES_set_key((const_DES_cblock*)key.data(), &keysched);
        //Encrypt Des
        encrypted_text = encryptDES(bytes, keysched);
    }
    
    

    //Convert to hexadecimal
    for (int i = 0; i < encrypted_text.size(); i++) {
        std::cout << std::hex << std::setfill('0') << std::setw(2) << (int)encrypted_text[i];
    }
}

File: 1_semestre/CA/Projeto_1/python/encrypt.py
from des import *
import sys


if __name__ == "__main__":
    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print("Usage: python encrypt.py <password> <mode:optional>")
        sys.exit(1)

    
    key = sys.argv[1]
    mode = "des"

    if len(sys.argv) == 3:
        mode = sys.argv[2]

    plaintext = sys.stdin.read()

    #Padder
    binary_plaintext = plaintext.encode()
    padder = padding.PKCS7(512).padder() # 64 bytes = 512 bit
    binary_plaintext= padder.update(binary_plaintext)
    binary_plaintext += padder.finalize()

    encrypted_text = ""
    #Encrypt
    if(mode == "edes"):
        #GenSbox
        sboxes=sbox_gen(random_gen(key,256))
        encrypted_text = encrypt_edes(binary_plaintext, sboxes)
    else:
        key = random_gen(key,8)
        cipher = DES.new(key, DES.MODE_ECB)
        encrypted_text = cipher.encrypt(binary_plaintext)
    

    sys.stdout.write(encrypted_text.hex())
